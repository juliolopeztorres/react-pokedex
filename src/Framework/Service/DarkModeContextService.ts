import { createContext } from 'react'

const DarkModeContextService = createContext<[boolean, (status: boolean) => void]>([false, () => {}])

export default DarkModeContextService
