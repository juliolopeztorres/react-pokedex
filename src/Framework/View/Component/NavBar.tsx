import React, { useContext } from "react";
import { hot } from "react-hot-loader/root";
import { DarkModeContext } from "./DarkModeContainer";
import { getRoute } from "../../Service/RouteService";
import { useHistory } from "react-router";

const NavBar = () => {
  const [darkMode, onDarkModeChanged] = useContext(DarkModeContext)
  const history = useHistory()

  return <div className='sticky z-10 top-0 flex flex-row bg-orange-600 shadow-md mb-4 pl-2 py-5 dark:bg-orange-800 dark:shadow-gray-700' style={{cursor: 'pointer'}}>
    <div className="grow flex flex-row">
      <div className="w-10 self-center">
        <img src="img/pokeball.png" className='w-7 mx-auto text-white font-mono dark:text-gray-200 text-xs' alt="pokeball"
             onClick={() => {
               history.push(getRoute('home'));
             }}/>
      </div>
      <span className='text-white text-2xl font-mono ml-2'>Pok√©dex</span>
    </div>
    <div className='mr-2'>
      <label className="relative inline-flex items-center cursor-pointer align-middle">
        <input type="checkbox" value="" className="sr-only peer" checked={darkMode} onChange={(e) => onDarkModeChanged!(e.target.checked)}/>
        <div className="w-11
            h-6
            bg-orange-200
            rounded-full
            peer
            peer-focus:outline-none
            peer-checked:after:translate-x-full
            peer-checked:after:border-white
            peer-checked:bg-orange-400
            after:content-['']
            after:absolute
            after:top-[2px]
            after:left-[2px]
            after:bg-orange-600
            after:rounded-full
            after:h-5 after:w-5 after:transition-all
            "/>
      </label>
    </div>
  </div>
}

export default hot(NavBar)
