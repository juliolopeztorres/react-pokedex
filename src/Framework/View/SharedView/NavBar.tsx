import React, { useContext } from "react";
import { hot } from "react-hot-loader/root";
import { useHistory } from "react-router";
import DarkModeContextService from "../../Service/DarkModeContextService";
import Pokeball from "../../Asset/Image/pokeball.png"
import getRoute from "../../../Domain/Util/getRoute";

const NavBar = () => {
  const [darkMode, onDarkModeChanged] = useContext(DarkModeContextService)
  const history = useHistory()

  return <div className='sticky top-0 flex flex-row bg-orange-600 shadow-md mb-4 pl-2 py-5 dark:bg-orange-800 dark:shadow-gray-700' style={{cursor: 'pointer'}}>
    <div className="grow flex flex-row">
      <div className="w-10 self-center">
        <img data-test-id='logo' src={Pokeball} className='w-7 mx-auto text-white font-mono dark:text-gray-200 text-xs' alt="pokeball"
             onClick={() => {
               history.push(getRoute('home'));
             }}/>
      </div>
      <span className='text-white text-2xl font-mono ml-2'>Pok√©dex</span>
    </div>
    <div className='mr-2'>
      <label className="relative inline-flex items-center cursor-pointer align-middle">
        <input data-test-id='checkbox' type="checkbox" value="" className="sr-only peer" checked={darkMode} onChange={(e) => onDarkModeChanged!(e.target.checked)}/>
        <div className="w-11
            h-6
            bg-orange-200
            rounded-full
            peer
            peer-focus:outline-none
            peer-checked:after:translate-x-full
            peer-checked:after:border-white
            peer-checked:bg-orange-400
            after:content-['']
            after:absolute
            after:top-[2px]
            after:left-[2px]
            after:bg-orange-600
            after:rounded-full
            after:h-5 after:w-5 after:transition-all
            "/>
      </label>
    </div>
  </div>
}

export default hot(NavBar)
