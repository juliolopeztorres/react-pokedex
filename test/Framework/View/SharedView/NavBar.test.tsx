import React from 'react'
import { createRoot, Root } from "react-dom/client"
import { act } from "@testing-library/react"

import DarkModeContextService from "../../../../src/Framework/Service/DarkModeContextService"
import { BrowserRouter } from "react-router-dom"

jest.mock('../../../../src/Framework/Asset/Image/pokeball.png', () => {
  return 'img/pokeball.png'
})

import NavBar from "../../../../src/Framework/View/SharedView/NavBar"

let root: Root
let container: HTMLElement
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div")
  document.body.appendChild(container)

  root = createRoot(container)
})

afterEach(() => {
  // cleanup on exiting
  act(() => root.unmount())

  container.remove()
})

it('should render', () => {
  const onDarkModeChanged = jest.fn()

  act(() => root.render(
    <BrowserRouter>
      <DarkModeContextService.Provider value={[false, onDarkModeChanged]}>
        <NavBar/>
      </DarkModeContextService.Provider>
    </BrowserRouter>
  ))

  expect(container.textContent).toContain('PokÃ©dex')

  expect(onDarkModeChanged).not.toHaveBeenCalled()
  const checkbox = container.querySelector('input[data-test-id="checkbox"]')
  expect(checkbox).toBeTruthy()

  act(() => checkbox!.dispatchEvent(new MouseEvent('click', {bubbles: true})))

  expect(onDarkModeChanged).toHaveBeenCalledTimes(1)
  expect(onDarkModeChanged).toHaveBeenLastCalledWith(true)

  expect(history.length).toBe(1)
  const img = container.querySelector('img[data-test-id="logo"]')
  expect(img).toBeTruthy()

  act(() => img!.dispatchEvent(new MouseEvent('click', {bubbles: true})))
  expect(history.length).toBe(2)

  expect(container).toMatchSnapshot()
})
