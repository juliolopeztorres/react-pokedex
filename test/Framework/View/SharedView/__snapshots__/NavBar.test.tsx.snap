// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div>
  <div
    class="sticky top-0 flex flex-row bg-orange-600 shadow-md mb-4 pl-2 py-5 dark:bg-orange-800 dark:shadow-gray-700"
    style="cursor: pointer;"
  >
    <div
      class="grow flex flex-row"
    >
      <div
        class="w-10 self-center"
      >
        <img
          alt="pokeball"
          class="w-7 mx-auto text-white font-mono dark:text-gray-200 text-xs"
          data-test-id="logo"
          src="img/pokeball.png"
        />
      </div>
      <span
        class="text-white text-2xl font-mono ml-2"
      >
        Pok√©dex
      </span>
    </div>
    <div
      class="mr-2"
    >
      <label
        class="relative inline-flex items-center cursor-pointer align-middle"
      >
        <input
          class="sr-only peer"
          data-test-id="checkbox"
          type="checkbox"
          value=""
        />
        <div
          class="w-11
            h-6
            bg-orange-200
            rounded-full
            peer
            peer-focus:outline-none
            peer-checked:after:translate-x-full
            peer-checked:after:border-white
            peer-checked:bg-orange-400
            after:content-['']
            after:absolute
            after:top-[2px]
            after:left-[2px]
            after:bg-orange-600
            after:rounded-full
            after:h-5 after:w-5 after:transition-all
            "
        />
      </label>
    </div>
  </div>
</div>
`
